File: legacy-node/lib/loyalty-service.js
const { createJsonStore, readJsonFile } = require("./persistence");

const DAILY_QUESTS = Object.freeze([
  Object.freeze({ id: "daily-claim", reason: "daily", label: "Claim the daily bonus", target: 1, reward: 5 }),
  Object.freeze({ id: "trivia-win", reason: "trivia", label: "Win a trivia question", target: 1, reward: 5 }),
  Object.freeze({ id: "marbles-entry", reason: "marbles", label: "Enter Marbles", target: 1, reward: 5 }),
  Object.freeze({ id: "event-join", reason: "event", label: "Join a community event", target: 1, reward: 5 }),
]);

function levelFor(points) {
  return 1 + Math.floor(Math.sqrt(Math.max(0, Number(points) || 0) / 25));
}

function levelThreshold(level) {
  return 25 * Math.max(0, level - 1) ** 2;
}

function titleFor(level) {
  if (level >= 10) return "Room Legend";
  if (level >= 7) return "Community Champion";
  if (level >= 5) return "Veteran";
  if (level >= 3) return "Regular";
  return "Newcomer";
}

class LoyaltyService {
  constructor({ filePath, debounceMs = 100, logger = console, database = null }) {
    const data = readJsonFile(filePath, {}, logger, database);
    this.users = new Map(Object.entries(data.users || {}));
    this.writer = createJsonStore({
      filePath,
      delayMs: debounceMs,
      label: "loyalty",
      logger,
      getData: () => ({ version: 1, users: Object.fromEntries(this.users) }),
      database,
    });
  }

  key(roomId, senderId, senderName) {
    return `${roomId}:${String(senderId || senderName || "unknown").toLowerCase()}`;
  }

  get(roomId, senderId, senderName) {
    const key = this.key(roomId, senderId, senderName);
    const user = this.users.get(key) || { roomId, id: senderId || null, name: senderName || "unknown", points: 0, triviaWins: 0, streak: 0, achievements: [], lastDaily: null };
    user.achievements ||= [];
    user.questDate ||= null;
    user.questProgress ||= {};
    user.completedQuests ||= [];
    return user;
  }

  recordQuest(user, reason) {
    const today = new Date().toISOString().slice(0, 10);
    if (user.questDate !== today) {
      user.questDate = today;
      user.questProgress = {};
      user.completedQuests = [];
    }
    const quest = DAILY_QUESTS.find((item) => item.reason === reason);
    if (!quest || user.completedQuests.includes(quest.id)) return null;
    user.questProgress[quest.id] = Math.min(quest.target, (user.questProgress[quest.id] || 0) + 1);
    if (user.questProgress[quest.id] < quest.target) return null;
    user.completedQuests.push(quest.id);
    user.points += quest.reward;
    return quest;
  }

  award(roomId, senderId, senderName, amount, reason = "participation") {
    const key = this.key(roomId, senderId, senderName);
    const user = this.get(roomId, senderId, senderName);
    user.name = senderName || user.name;
    user.points += Number(amount) || 0;
    if (reason === "trivia") user.triviaWins += 1;
    const completedQuest = this.recordQuest(user, reason);
    const unlocked = [];
    for (const achievement of [
      { id: "first-point", threshold: 1, label: "First Point" },
      { id: "regular", threshold: 50, label: "Room Regular" },
      { id: "centurion", threshold: 100, label: "Centurion" },
      { id: "trivia-ace", test: () => user.triviaWins >= 10, label: "Trivia Ace" },
    ]) {
      const earned = achievement.test ? achievement.test() : user.points >= achievement.threshold;
      if (earned && !user.achievements.includes(achievement.id)) {
        user.achievements.push(achievement.id);
        unlocked.push(achievement.label);
      }
    }
    this.users.set(key, user);
    this.writer.schedule();
    return { user, unlocked, completedQuest };
  }

  daily(roomId, senderId, senderName) {
    const user = this.get(roomId, senderId, senderName);
    const today = new Date().toISOString().slice(0, 10);
    if (user.lastDaily === today) return { ok: false, user };
    user.lastDaily = today;
    this.users.set(this.key(roomId, senderId, senderName), user);
    const result = this.award(roomId, senderId, senderName, 10, "daily");
    return { ok: true, ...result };
  }

  leaderboard(roomId, limit = 5) {
    return [...this.users.values()]
      .filter((user) => String(user.roomId) === String(roomId))
      .sort((a, b) => b.points - a.points)
      .slice(0, limit);
  }

  progress(roomId, senderId, senderName) {
    const user = this.get(roomId, senderId, senderName);
    const level = levelFor(user.points);
    return {
      user,
      level,
      title: titleFor(level),
      currentLevelAt: levelThreshold(level),
      nextLevelAt: levelThreshold(level + 1),
    };
  }

  quests(roomId, senderId, senderName) {
    const user = this.get(roomId, senderId, senderName);
    this.recordQuest(user, "__refresh__");
    return DAILY_QUESTS.map((quest) => ({
      ...quest,
      progress: user.questProgress[quest.id] || 0,
      completed: user.completedQuests.includes(quest.id),
    }));
  }

  flush() { return this.writer.flush(); }
}

module.exports = { DAILY_QUESTS, LoyaltyService, levelFor, levelThreshold, titleFor };

File: Services/LoyaltyService.cs
using EpikChatBot.Core;

namespace EpikChatBot.Services;

public sealed record LoyaltyUser { public string Id { get; set; } = ""; public string Name { get; set; } = ""; public int Points { get; set; } public int TriviaWins { get; set; } public List<string> Achievements { get; set; } = []; public string LastDaily { get; set; } = ""; public Dictionary<string, int> DailyProgress { get; set; } = []; }
public sealed record Quest(string Id, string Label, int Target, int Reward, int Progress, bool Completed);

public sealed class LoyaltyService
{
  private readonly string _filePath; private readonly Dictionary<string, Dictionary<string, LoyaltyUser>> _rooms;
  public LoyaltyService(string filePath)
  {
    _filePath = filePath; _rooms = [];
    if (File.Exists(filePath)) try { using var document = System.Text.Json.JsonDocument.Parse(File.ReadAllText(filePath)); var users = document.RootElement.TryGetProperty("users", out var nested) ? nested : document.RootElement; foreach (var entry in users.EnumerateObject()) { var split = entry.Name.Split(':', 2); if (split.Length != 2) continue; if (!_rooms.TryGetValue(split[0], out var room)) _rooms[split[0]] = room = []; room[split[1]] = System.Text.Json.JsonSerializer.Deserialize<LoyaltyUser>(entry.Value, JsonStore.Options) ?? new LoyaltyUser { Id = split[1], Name = split[1] }; } } catch (Exception exception) { JsonStore.QuarantineCorruptFile(filePath, exception); }
  }
  public LoyaltyUser Get(string room, string? id, string? name)
  {
    if (!_rooms.TryGetValue(room, out var users)) _rooms[room] = users = []; var key = (id ?? name ?? "unknown").ToLowerInvariant();
    if (!users.TryGetValue(key, out var user)) users[key] = user = new() { Id = key, Name = name ?? id ?? "friend" }; if (!string.IsNullOrWhiteSpace(name)) user.Name = name; return user;
  }
  public LoyaltyUser Award(string room, string? id, string? name, int points, string source)
  {
    var user = Get(room, id, name); user.Points += points; if (source == "trivia") user.TriviaWins++; if (!user.Achievements.Contains("first-point")) user.Achievements.Add("first-point");
    var day = DateTime.UtcNow.ToString("yyyy-MM-dd"); var progressKey = $"{day}:{source}"; user.DailyProgress[progressKey] = user.DailyProgress.GetValueOrDefault(progressKey) + 1;
    var quest = Quests(room, id, name).FirstOrDefault(x => x.Id == $"{source}-win" && x.Completed); if (quest is not null && user.DailyProgress.GetValueOrDefault($"{day}:reward:{quest.Id}") == 0) { user.Points += quest.Reward; user.DailyProgress[$"{day}:reward:{quest.Id}"] = 1; }
    return user;
  }
  public (bool Ok, LoyaltyUser User) Daily(string room, string? id, string? name)
  {
    var user = Get(room, id, name); var day = DateTime.UtcNow.ToString("yyyy-MM-dd"); if (user.LastDaily == day) return (false, user);
    user.LastDaily = day; user.Points += 10; user.DailyProgress[$"{day}:daily"] = 1; user.Points += 5; user.DailyProgress[$"{day}:reward:daily-claim"] = 1; if (!user.Achievements.Contains("first-point")) user.Achievements.Add("first-point"); return (true, user);
  }
  public IReadOnlyList<Quest> Quests(string room, string? id, string? name)
  {
    var user = Get(room, id, name); var day = DateTime.UtcNow.ToString("yyyy-MM-dd");
    return [new("daily-claim", "Claim the daily bonus", 1, 5, user.DailyProgress.GetValueOrDefault($"{day}:daily"), user.LastDaily == day), new("trivia-win", "Win a trivia question", 1, 10, user.DailyProgress.GetValueOrDefault($"{day}:trivia"), user.DailyProgress.GetValueOrDefault($"{day}:trivia") >= 1), new("marbles-win", "Join Marbles", 1, 5, user.DailyProgress.GetValueOrDefault($"{day}:marbles"), user.DailyProgress.GetValueOrDefault($"{day}:marbles") >= 1), new("event-win", "Join an event", 1, 5, user.DailyProgress.GetValueOrDefault($"{day}:event"), user.DailyProgress.GetValueOrDefault($"{day}:event") >= 1)];
  }
  public static int LevelFor(int points) => Math.Max(1, (int)Math.Floor(Math.Sqrt(Math.Max(0, points) / 25d)) + 1);
  public static string TitleFor(int level) => level switch { >= 10 => "Legend", >= 7 => "Champion", >= 5 => "Veteran", >= 3 => "Regular", _ => "Newcomer" };
  public (LoyaltyUser User, int Level, string Title, int NextLevelAt) Progress(string room, string? id, string? name) { var user = Get(room, id, name); var level = LevelFor(user.Points); return (user, level, TitleFor(level), level * level * 25); }
  public IReadOnlyList<LoyaltyUser> Leaderboard(string room, int limit = 10) => _rooms.GetValueOrDefault(room)?.Values.OrderByDescending(x => x.Points).Take(limit).ToList() ?? [];
  public Task FlushAsync() => JsonStore.WriteAsync(_filePath, new { version = 1, users = _rooms.SelectMany(room => room.Value.Select(user => new KeyValuePair<string, LoyaltyUser>($"{room.Key}:{user.Key}", user.Value))).ToDictionary() });
}
